### PR TITLE
RFC: Fix #13957 by refactoring _svds

### LINKED ISSUE #13957: RFC: Improve svds
- [ ] The first 3 output quantities `(left_sv, s, right_sv)` could be put into an `SVD` object if `ritzvec = true`, so that the number of outputs from `svds` would always be the same, with the type of the first output being a `Vector` of singular values if `ritzvec=false` and a `SVD` if `ritzvec=true`. (In reality the `SVD` is now representative of an incomplete singular value decomposition rather than a full calculation, but I think the chance of confusion is low.)

- [ ] The current docs say that the 6th positional output is `nmult`, but it does not explain what a "mult" is in this context. The documentation should explain that a "mult" is a single evaluation of the `A'A` operator, since it's no longer clear which of A'A*v, A'v or A*v is the product in question.

- [ ] `svds` should accept a `ncv` kwarg, which controls the maximum size of Krylov subspace used, and is passed directly to the underlying call to `eigs` anyway.

- [ ] `svds` should accept a `v0` kwarg to specify the initial Krylov vector; ref #10330

### PR DESCRIPTION
**Pull request by kshyatt**
Added `ncv` and `v0` kwargs. Redid some documentation. Output is now a
SVD object in either `ritzvec` case, with empty/non-empty singular
vectors.

### CODE REVIEW

**Comment by kshyatt** [comment]
@jiahao @andreasnoack @tkelman relevant to your interests.

File: base/linalg/arnoldi.jl [hunk 1]
```diff
@@ -307,7 +307,7 @@ function svds{T}(A::AbstractMatrix{T}; kwargs...)
 end
 
 """
-    svds(A; nsv=6, ritzvec=true, tol=0.0, maxiter=1000) -> ([left_sv,] s, [right_sv,] nconv, niter, nmult, resid)
+    svds(A; nsv=6, ritzvec=true, tol=0.0, maxiter=1000, ncv=2*nsv, v0=zeros((0,))) -> (SVD([left_sv,] s, [right_sv,]), nconv, niter, nmult, resid)
```
Comment (tkelman) [inline base/linalg/arnoldi.jl]: update the rst docs

**Comment by andreasnoack** [comment]
Great. Thanks for fixing this. I also think it is a good idea to wrap the result in the `SVD` type to make the function type stable.

**Comment by jiahao** [comment]
Thanks!

Could you put some version of https://github.com/JuliaLang/julia/issues/10330#issuecomment-137597309 into the documentation for `v0`? It's actually rather tricky to work out how exactly to specify the starting guess vector.

**Comment by kshyatt** [comment]
@jiahao I added the documentation and I added bounds checking for `v0` since there are only 4 allowed lengths.

File: base/linalg/arnoldi.jl [hunk 2]
```diff
@@ -349,23 +349,36 @@ Lanczos tridiagonalization on the Hermitian matrix
...
         throw(ArgumentError("number of singular values (nsv) must be ≥ 1, got $nsv"))
     end
     if nsv > minimum(size(X))
         throw(ArgumentError("number of singular values (nsv) must be ≤ $(minimum(size(X))), got $nsv"))
     end
+    m,n = size(X)
     otype = eltype(X)
-    ex    = eigs(SVDOperator(X), I; ritzvec = ritzvec, nev = 2*nsv, tol = tol, maxiter = maxiter)
-    ind   = [1:2:nsv*2;]
+    if length(v0) == n
```
Comment (tkelman) [inline base/linalg/arnoldi.jl]: which is a better choice in the square case?

File: base/linalg/arnoldi.jl [hunk 3]
```diff
@@ -349,23 +349,36 @@ Lanczos tridiagonalization on the Hermitian matrix
...
-    right_sv = sqrt(2) * ex[2][ size(X,1)+1:end, ind ]
-    return (left_sv, sval, right_sv, ex[3], ex[4], ex[5], ex[6])
+    if ritzvec
+        # calculating singular vectors
+        left_sv  = sqrt(2) * ex[2][ 1:size(X,1),     ind ] .* sign(ex[1][ind]')
+        right_sv = sqrt(2) * ex[2][ size(X,1)+1:end, ind ]
+        return (SVD(left_sv, sval, right_sv), ex[3], ex[4], ex[5], ex[6])
+    else
+        #The sort is necessary to work around #10329
+        m,n = size(X)
```
Comment (tkelman) [inline base/linalg/arnoldi.jl]: it hasn't changed size by here, has it?

Comment (kshyatt) [inline base/linalg/arnoldi.jl]: No, it hasn't.

File: base/linalg/arnoldi.jl (hunk 2 above)
Comment (kshyatt) [inline base/linalg/arnoldi.jl]: @jiahao or @andreasnoack?

**Comment by ViralBShah** [comment]
Great PR. Unrelated here, there is perhaps potential for higher performance by reusing memory in the matvec rather than allocating a new vector every iteration. Since you are deep in this code, thought I would mention it.

File: base/linalg/arnoldi.jl (hunk 3 above)
Comment (tkelman) [inline base/linalg/arnoldi.jl]: then this line is unnecessary, yeah?

File: base/linalg/arnoldi.jl (hunk 2 above)
Comment (jiahao) [inline base/linalg/arnoldi.jl]: there is no 'better' case, they are merely different

Comment (tkelman) [inline base/linalg/arnoldi.jl]: It's ambiguous which way the user would like to initialize then. Would it make sense to be strict and require complete initialization, or provide some different mechanism to indicate the intent of which part of the vector should be initialized with the user provided values?

File: base/linalg/arnoldi.jl [hunk 4]
```diff
@@ -318,17 +318,18 @@ iterations derived from [`eigs`](:func:`eigs`).
...
-* `nsv`: Number of singular values.
+* `nsv`: Number of singular values. Default: 6.
 * `ritzvec`: If `true`, return the left and right singular vectors `left_sv` and `right_sv`.
    If `false`, omit the singular vectors. Default: `true`.
 * `tol`: tolerance, see [`eigs`](:func:`eigs`).
-* `maxiter`: Maximum number of iterations, see [`eigs`](:func:`eigs`).
+* `maxiter`: Maximum number of iterations, see [`eigs`](:func:`eigs`). Default: 1000.
+* `ncv`: Maximum size of the Krylov subspace, see [`eigs`](:func:`eigs`) (there called `nev`). Default: `2*nsv`.
+* `u0`: Initial guess for the first left Krylov vector. It may have length `m`, the first dimension of `A`.
+* `v0`: Initial guess for the first right Krylov vector. It may have length `n`, the second dimension of `A`.
```
Comment (tkelman) [inline base/linalg/arnoldi.jl]: what do you mean by "may have length" ?

Comment (kshyatt) [inline base/linalg/arnoldi.jl]: I mean that that is the allowed length.

Comment (tkelman) [inline base/linalg/arnoldi.jl]: should? or looking lower in the file, looks like it's allowed to be either empty or this specific size but nothing else

Comment (kshyatt) [inline base/linalg/arnoldi.jl]: I'm updating the docs to make that more clear, adding tests, and fixing the whitespace. Stand by for rebase.

**Comment by kshyatt** [comment]
Rebased with more tests & better docs.

File: base/linalg/arnoldi.jl [hunk 5]
```diff
@@ -318,17 +318,18 @@ iterations derived from [`eigs`](:func:`eigs`).
...
-* `nsv`: Number of singular values.
+* `nsv`: Number of singular values. Default: 6.
 * `ritzvec`: If `true`, return the left and right singular vectors `left_sv` and `right_sv`.
    If `false`, omit the singular vectors. Default: `true`.
 * `tol`: tolerance, see [`eigs`](:func:`eigs`).
-* `maxiter`: Maximum number of iterations, see [`eigs`](:func:`eigs`).
+* `maxiter`: Maximum number of iterations, see [`eigs`](:func:`eigs`). Default: 1000.
+* `ncv`: Maximum size of the Krylov subspace, see [`eigs`](:func:`eigs`) (there called `nev`). Default: `2*nsv`.
+* `u0`: Initial guess for the first left Krylov vector. It may have length `m`, the first dimension of `A`, or 0.
+* `v0`: Initial guess for the first right Krylov vector. It may have length `n`, the second dimension of `A`, or 0.
```
Comment (tkelman) [inline base/linalg/arnoldi.jl]: Needs a `make docs` sync and/or tweaking the rst signature to match again? The commas here make it almost sound like this is a list of 3 allowed lengths, but it's close enough.

Comment (kshyatt) [inline base/linalg/arnoldi.jl]: I'll get rid of the ambiguity and do make-docs. Sorry. I ran `genstdlib.jl` locally too...

**Comment by kshyatt** [comment]
This is just awful. I can't do a reaction emoji directly onto the act of merging? Come on. 🎆 🎉 🍺

**Comment by madeleineudell** [comment]
The output of `svds` is now inconsistent with the output of `svd`. Is there a reason for this?

**Comment by andreasnoack** [comment]
I'm not sure what you mean with "inconsistent"? Can you be more specific?

**Comment by madeleineudell** [comment]
`svds(A)` returns an SVD object, whereas `svd(A)` returns a tuple of U, S, and V with U*diagm(S)*V' = A.

```
julia> A = rand(3,4);

julia> svd(A)
(
[-0.440632 0.262519 -0.858444; -0.574102 0.652746 0.494297; -0.690109 -0.710638 0.136908],

[1.72609,0.933311,0.442869],
[-0.60164 -0.34264 0.0897604; -0.565715 -0.452677 -0.0264048; -0.427674 0.671261 0.59481; -0.367547 0.476539 -0.798403])

julia> svds(A, nsv=3)
(Base.LinAlg.SVD{Float64,Float64,Array{Float64,2}}([0.440632 -0.262519 -0.858444; 0.574102 -0.652746 0.494297; 0.690109 0.710638 0.136908],[1.72609,0.933311,0.442869],[0.60164 0.34264 0.0897604; 0.565715 0.452677 -0.0264048; 0.427674 -0.671261 0.59481; 0.367547 -0.476539 -0.798403]),6,1,7,[0.0,0.0,0.0,0.0,0.0,0.0,0.0])
```

**Comment by andreasnoack** [comment]
Got it. First It thought there was a problem with the numerical output from `svds`. I think the idea is that the `SVD` type is a convenient way of collecting the vectors and values of the SVD. It is also what we return from `svdfact`. It used to be the case that `svds` returned output tuples of varying length depending on the value of `ritzvec`. Now it always returns the same type. Except for this being a breaking change, is it a problem to work with the `SVD` type?

**Comment by madeleineudell** [comment]
I agree the SVDFact is a fine type to work with. But to me it seems very
strange for svd and svds to return different types; I often swap out one
for the other as I scale up my experiments. And it's particularly
surprisingly that this change of syntax was worth a breaking change.

On Oct 29, 2016 7:53 AM, "Andreas Noack" <notifications@github.com> wrote:

Got it. First It thought there was a problem with the numerical output from
svds. I think the idea is that the SVD type is a convenient way of
collecting the vectors and values of the SVD. It is also what we return
from svdfact. It used to be the case that svds returned output tuples of
varying length depending on the value of ritzvec. Now it always returns the
same type. Except for this being a breaking change, is it a problem to work
with the SVD type?

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/JuliaLang/julia/pull/16283#issuecomment-257087454>, or mute
the thread

**Comment by madeleineudell** [comment]
I agree the SVDFact is a fine type to work with. But to me it seems very
strange for svd and svds to return different types; I often swap out one
for the other as I scale up my experiments. And it's particularly
surprisingly that this change of syntax was worth a breaking change.

**Comment by andreasnoack** [comment]
> But to me it seems very strange for svd and svds to return different types

They already did that before this change because 1) of the `ritzvec` keyword argument 2) the extra output related to the Lanczos process. I think achieving type stability is worth a breaking change until we reach 1.0. We could have achieved type stability in other ways but nobody suggested alternatives at the time we changed this.

**Pull request merged**